R/ZIPLNfit-class.R

```diff
@@ -0,0 +1,831 @@
+#' @param covariates design matrix (called X in the model). Will usually be extracted from the corresponding field in PLNfamily-class
+#' @param offsets offset matrix (called O in the model). Will usually be extracted from the corresponding field in PLNfamily-class
+#' @param data an optional data frame, list or environment (or object coercible by as.data.frame to a data frame) containing the variables in the model. If not found in data, the variables are taken from environment(formula), typically the environment from which PLN is called.
+#' @param weights an optional vector of observation weights to be used in the fitting process.
+#' @param nullModel null model used for approximate R2 computations. Defaults to a GLM model with same design matrix but not latent variable.
+#' @param formula model formula used for fitting, extracted from the formula in the upper-level call
+#' @param control a list for controlling the optimization. See details.
+#' @param config part of the \code{control} argument which configures the optimizer
+#'
+#' @importFrom R6 R6Class
+#'
+#' @examples
+#' \dontrun{
+#' # See other examples in function ZIPLN
+#' data(trichoptera)
+#' trichoptera <- prepare_data(trichoptera$Abundance, trichoptera$Covariate)
+#' myPLN <- ZIPLN(Abundance ~ 1, data = trichoptera, control=  ZIPLN_param(penalty = 0.2))
+#' class(myPLN)
+#' print(myPLN)
+#' }
+ZIPLNfit_sparse <- R6Class(
+  classname = "ZIPLNfit_sparse",
+  inherit = ZIPLNfit,
+  ## %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+  ## PUBLIC MEMBERS ----
+  ## %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+  public  = list(
+    #' @description Initialize a [`ZIPLNfit_fixed`] model
+    #' @importFrom glassoFast glassoFast
+    initialize = function(responses, covariates, offsets, weights, formula, control) {
+      super$initialize(responses, covariates, offsets, weights, formula, control)
+      private$optimizer$Omega <-
+        function(M, X, B, S) {
+          glassoFast( crossprod(M - X %*% B)/self$n + diag(colMeans(S * S), self$p, self$p), rho = control$penalty )$wi
+        }
+    }
+  ),
+  active = list(
+    #' @field nb_param number of parameters in the current PLN model
+    nb_param   = function() {
+      res <-  self$p * self$d + (sum(private$Omega != 0) - self$p)/2L +
+        switch(private$ziparam,
+               "single" = 1,
+               "row"    = self$n,
+               "col"    =  self$p,
+               "covar"  =  self$p * self$d)
+      as.integer(res)
+    },
+    #' @field vcov_model character: the model used for the residual covariance
+    vcov_model = function() {"sparse"}
+  )
+  ## %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+  ##  END OF THE CLASS ZIPLNfit_sparse
+  ## %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+)
+
+
+# Test convergence for a named list of parameters
```

Comment:
To move to utils-zipln.R ?

Reply:
Good idea,will be more relevant/readable.